add nbt command

LordMZTE opened the pull request.

LordMZTE committed:
add nbt command

LordMZTE commented:
![image](https://user-images.githubusercontent.com/28735087/100545324-9b121b80-325b-11eb-99b2-3ffa622e5f39.png)
includes fancy UI!

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/gui/InteractableScreen.kt:
```diff
@@ -0,0 +1,50 @@
+ * class for a screen for which input should be handled.
+ * includes callback handlers
+ */
+abstract class InteractableScreen(title: Text) : Screen(title) {
```
This name seems to imply a more generic "interaction" class, but it actually implements dear imgui (something specific).
The name should reflect that - e.g. `ImGuiScreen`

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+    itemWidgetsUtilities,
+    columns,
+    cursorLayout
+// this comment is here to prevent intelli'ntJ from formatting this against the kotlin conventions
```
Now this is the proper way to give me a heart attack - we are *not* supposed to be implementing these interfaces!!

`ImGui` implements them, and all of your methods should be using calling `ImGui`'s methods and not `this`.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    operator fun invoke() {
```
```suggestion
    operator fun invoke() = with(ImGui) {

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    private fun showTree(tagName: String, tag: Tag, appendix: Int = 0) {
```
```suggestion
    private fun showTree(tagName: String, tag: Tag, appendix: Int = 0) = with(ImGui) {
```

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+        else -> defaultColor
+    }
+
+    private fun nbtLabelText(k: String, v: String) {
```
```suggestion
    private fun nbtLabelText(k: String, v: String): Unit = with(ImGui) {
```
(The `:Unit` is because kotlin type checking will otherwise fail in the recursive call)

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+
+class NbtScreen(
+    var tag: Tag,
+    val defaultColor: Vec4 = Vec4(1f, 1f, 1f, 1f)
```
```suggestion
    private val defaultColour: Vec4 = Vec4(1f, 1f, 1f, 1f)
```
:gb:

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+                        HitResult.Type.BLOCK -> {
+                            val entity = mc.world?.getBlockEntity(BlockPos(target.pos))
+                            if (entity == null) {
+                                Error.NO_BLOCK_ENTITY.send(it.source)
```
Errors in brigadier are supposed to be thrown exceptions. Brigadier will handle them differently and show the user a formatted error.

Take a look at `PeekCommand`'s `FAILED_EXCEPTION` and when it is thrown.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+                        }
+                        HitResult.Type.ENTITY -> mc.targetedEntity?.toTag(CompoundTag())
+                        else -> {
+                            Error.NO_TARGET.send(it.source)
```
Error: ditto

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+                does {
+                    val stack = mc.player?.mainHandStack
+                    if (stack?.isEmpty == true) {
+                        Error.NO_ITEM.send(it.source)
```
Error: ditto

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+                        return@does 1
+                    } else {
+                        val tag = stack?.tag
+                            ?: run { Error.NO_ITEM_NBT.send(it.source); return@does 1 }
```
Error: ditto

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+    cursorLayout
+// this comment is here to prevent intelli'ntJ from formatting this against the kotlin conventions
+{
+    override fun render(matrices: MatrixStack?, mouseX: Int, mouseY: Int, delta: Float) {
+        super.render(matrices, mouseX, mouseY, delta)
+
+        KamiHud.frame(matrices!!) {
```
```suggestion
    override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
        super.render(matrices, mouseX, mouseY, delta)

        KamiHud.frame(matrices) {
```

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+        dsl.window("NBT") {
+            if (button("Copy", Vec2(windowContentRegionWidth, 0)))
+                copyTagToClipboard(tag)
+            beginChild("##nbtPane", border = true)
+            columns(2)
+            showTree("NBT", tag)
+            columns(1)
+            endChild()
```
```suggestion
            child("##nbtPane", border = true) {
                columns(2)
                showTree("NBT", tag)
                columns(1)
            }
```
`child` comes from `imgui.dsl`.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+            beginChild("##nbtPane", border = true)
+            columns(2)
+            showTree("NBT", tag)
+            columns(1)
```
```suggestion
            columns()
```

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+                            if (tagIsDeep(it))
+                                showTree(key, it, ++curApp)
+                            else
+                                dsl.withStyleColor(Col.Text, tagColor(it)) {
```
```suggestion
                                withStyleColor(Col.Text, tagColor(it)) {
```
Add an import for `withStyleColor`.

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+                        if (tagIsDeep(t))
+                            showTree("[array entry]", t, ++curApp)
+                        else
+                            dsl.withStyleColor(Col.Text, tagColor(t)) {
```
```suggestion
                            withStyleColor(Col.Text, tagColor(t)) {
```

zeroeightysix commented on src/main/kotlin/me/zeroeightsix/kami/feature/command/NbtCommand.kt:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    operator fun invoke() {
+        dsl.window("NBT") {
```
```suggestion
        window("NBT") {
```
Add an import for `dsl.window`.

zeroeightysix reviewed (changes requested):
This looks great! A few minor code tweaks needed, but the feature itself is amazing.

LordMZTE committed:
implement requested changes to nbt gui
rename InteractableScreen to ImGuiScreen
use `with` instead of implementing a lot of interfaces

LordMZTE committed:
implement rest of requested nbt command changes
make defaultColor private
use proper command error
remove pointless null check ignore

zeroeightysix commented:
Thanks!
